### PR TITLE
Change SocketCache to be shared between nodes

### PR DESCRIPTION
**Pull request by ganmacs**
**Which issue(s) this PR fixes**: 

no

**What this PR does / why we need it**: 

Old SocketCache isn't shared between other nodes since node's `port` and `host` is fixed at start time. 
In the situation nodes' port and host change dynamically, sharing SocketCache between nodes makes more efficient. 
And recent refactoring can make SocketCache more simple.

**Docs Changes**:

no need

**Release Note**: 

no need

### CODE REVIEW

File: lib/fluent/plugin/out_forward/socket_cache.rb [hunk 1]
```diff
@@ -19,135 +19,118 @@
...
+
+            @inflight_sockets[val] = new_tsock
+            return new_tsock.sock
           end
+          tsock.sock
         end
       end
 
-      def clear
+      def checkin(sock)
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward/socket_cache.rb]: Hmm... `checkin` is unclear for me. After `checkin`, we are inflight for airplane but this method changes inflight to available. `checkout` too. Maybe, we don't use `checkout` for flight.

File: lib/fluent/plugin/out_forward/socket_cache.rb [hunk 2]
```diff
@@ -19,135 +19,118 @@
...
-          @active_socks.keys.each do |k|
-            if expired?(k) && @active_socks[k].ref <= 0
-              @inactive_socks[k] = @active_socks.delete(k)
-            end
-          end
+          sockets += @inactive_sockets
+          @inactive_sockets.clear
+        end
+
+        while (s = sockets.pop)
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward/socket_cache.rb]: `each` is more lightweight due to no `sockets` update.

File: lib/fluent/plugin/out_forward/socket_cache.rb [hunk 3]
```diff
@@ -19,135 +19,118 @@
...
-      def dec_ref(key = Thread.current.object_id)
-        @mutex.synchronize do
-          if @active_socks[key]
-            @active_socks[key].ref -= 1
-          elsif @inactive_socks[key]
-            @inactive_socks[key].ref -= 1
-          else
-            @log.warn("Not found key for dec_ref: #{key}")
-          end
+        while (s = sockets.pop)
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward/socket_cache.rb]: ditto

File: lib/fluent/plugin/out_forward/socket_cache.rb [hunk 4]
```diff
@@ -19,135 +19,118 @@
...
-            @inactive_socks[key].ref = 0
+          tsock = pick_socket(key)
+
+          unless tsock
+            val = yield
+            new_tsock = TimedSocket.new(timeout, key, val)
+            @log.debug("connect new socket #{new_tsock}")
+
+            @inflight_sockets[val] = new_tsock
+            return new_tsock.sock
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward/socket_cache.rb]: To reduce `return` by `if`

```
if tsock
  tsock.sock
else
  # same
end
```

File: lib/fluent/plugin/out_forward/socket_cache.rb (hunk 1 above)
Comment (ganmacs) [reply lib/fluent/plugin/out_forward/socket_cache.rb]: I referred connection_pool gem to determine the name of them.
https://github.com/mperham/connection_pool/blob/99b81aa0a9dccb3b423260740ce2d603f1479356/lib/connection_pool.rb#L80

Any name is okay for me if it's clear for readers. Do you have any suggestions?

File: lib/fluent/plugin/out_forward/socket_cache.rb (hunk 4 above)
Comment (ganmacs) [reply lib/fluent/plugin/out_forward/socket_cache.rb]: fixed https://github.com/fluent/fluentd/pull/2501/commits/5dd10e6f630f3b90f4911798387d144c1411e9bf

File: lib/fluent/plugin/out_forward/socket_cache.rb (hunk 2 above)
Comment (ganmacs) [reply lib/fluent/plugin/out_forward/socket_cache.rb]: right……… fixed https://github.com/fluent/fluentd/pull/2501/commits/eda267cee62e3230d700037fef0fd90a3076c5fa

File: lib/fluent/plugin/out_forward/socket_cache.rb (hunk 1 above)
Comment (repeatedly) [reply lib/fluent/plugin/out_forward/socket_cache.rb]: Hmm.... I see. If this naming is popular in ruby, it is okay for me.

File: lib/fluent/plugin/out_forward/socket_cache.rb [hunk 5]
```diff
@@ -19,135 +19,119 @@
...
         @mutex.synchronize do
-          if @active_socks[key]
-            @inactive_socks[key] = @active_socks.delete(key)
-            @inactive_socks[key].ref = 0
+          tsock = pick_socket(key)
+
+          if tsock
+            tsock.sock
+          else
+            val = yield
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward/socket_cache.rb]: `sock` or `new_sock` is better than `val`

Comment (ganmacs) [reply lib/fluent/plugin/out_forward/socket_cache.rb]: fixed https://github.com/fluent/fluentd/pull/2501/commits/14985b352e909c984f26e44253bcdff13116a0ff

**Comment by repeatedly** [comment]
looks good.

**Pull request merged**